evolver/calibration/standard/actions/temperature.py

```diff
@@ -1,26 +1,25 @@
-from typing import Dict, Optional
+from typing import Optional
 
 from pydantic import BaseModel, Field
 
 from evolver.calibration.action import CalibrationAction
+from evolver.calibration.interface import CalibrationStateModel
 
 
 class ReferenceValueAction(CalibrationAction):
     class FormModel(BaseModel):
-        """
-        Because this action requires manual user input, we define a Pydantic model for the input to the action.
-        This is used by the frontend to generate a form for the user to fill out.
-        """
-
         temperature: float = Field(..., title="Temperature", description="Temperature in degrees Celsius")
 
     def __init__(self, vial_idx: int, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.vial_idx = vial_idx
 
-    def execute(self, state: Dict, payload: Optional[FormModel] = None):
-        state.setdefault(self.vial_idx, {"reference": [], "raw": []})
-        state[self.vial_idx]["reference"].append(payload.temperature)
+    def execute(self, state: CalibrationStateModel, payload: Optional[FormModel] = None):
+        if not state.measured:
+            state.measured = {}
+        if self.vial_idx not in state.measured:
+            state.measured[self.vial_idx] = {"reference": [], "raw": []}
+        state.measured[self.vial_idx]["reference"].append(payload.temperature)
```

Comment:
totally fine, but in case you like: `collections.defaultdict` and/or `setdefault` is useful here:

```py
state.measured = state.measured or defaultdict(lambda: {"referece": [], "raw": []})
state.measured[self.vial_idx]["reference"].append(payload.temperature)
```

where all elements are same kind, no need to check and initialize yourself.